File: liva-sdk-ios/LIVAAnimation/Sources/Audio/AudioSyncManager.swift
//
//  AudioSyncManager.swift
//  LIVAAnimation
//
//  Coordinates audio playback with animation timing.
//

import Foundation

/// Synchronizes audio playback with frame rendering
final class AudioSyncManager {

    // MARK: - Properties

    private weak var audioPlayer: AudioPlayer?
    private weak var animationEngine: AnimationEngine?

    /// Current playback position in frames
    private var currentFrame: Int = 0

    /// Target frames per audio chunk
    private let framesPerChunk: Int = 45

    // MARK: - Initialization

    init(audioPlayer: AudioPlayer, animationEngine: AnimationEngine) {
        self.audioPlayer = audioPlayer
        self.animationEngine = animationEngine
    }

    // MARK: - Synchronization

    /// Start synchronized playback
    func startSync() {
        currentFrame = 0
        animationEngine?.setMode(.talking)
    }

    /// Handle chunk completion
    func onChunkComplete(chunkIndex: Int) {
        // Calculate expected frame position
        let expectedFrame = (chunkIndex + 1) * framesPerChunk

        // Adjust animation timing if needed
        // TODO: Implement sync adjustment
    }

    /// Stop synchronized playback
    func stopSync() {
        animationEngine?.setMode(.idle)
        currentFrame = 0
    }

    /// Called when audio ends
    func onAudioEnd() {
        // Transition to idle after short delay
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
            self?.animationEngine?.setMode(.idle)
        }
    }
}

File: liva-sdk-android/liva-animation/src/main/kotlin/com/liva/animation/audio/AudioSyncManager.kt
// @know entity AudioSyncManager_Android
package com.liva.animation.audio

import com.liva.animation.rendering.AnimationEngine
import com.liva.animation.rendering.AnimationMode
import kotlinx.coroutines.*

/**
 * Synchronizes audio playback with frame rendering.
 */
internal class AudioSyncManager(
    private val audioPlayer: AudioPlayer,
    private val animationEngine: AnimationEngine
) {
    private var currentFrame = 0
    private val framesPerChunk = 45

    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    fun startSync() {
        currentFrame = 0
        animationEngine.setMode(AnimationMode.TALKING)
    }

    fun onChunkComplete(chunkIndex: Int) {
        val expectedFrame = (chunkIndex + 1) * framesPerChunk
        // TODO: Implement sync adjustment
    }

    fun stopSync() {
        animationEngine.setMode(AnimationMode.IDLE)
        currentFrame = 0
    }

    fun onAudioEnd() {
        scope.launch {
            delay(500)
            animationEngine.setMode(AnimationMode.IDLE)
        }
    }

    fun cancel() {
        scope.cancel()
    }
}
